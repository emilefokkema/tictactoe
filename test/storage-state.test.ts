import { describe, beforeEach, it, expect } from 'vitest'
import { StorageState } from '../src/scripts/store/storage-state';
import winner01 from './fixtures/winner-0-1.json'
import manyLosers from './fixtures/many-losers.json'
import { replayRecord } from './replay-record';
import { GameState } from '../src/scripts/state/game-state';
import { RevealedPosition } from '../src/scripts/state/revealed-position';

function serializeAndDeserialize(state: StorageState): StorageState {
    const serialized = JSON.stringify(state);
    const deserialized = StorageState.fromJSON(JSON.parse(serialized));
    return deserialized;
}

function restoreRevealedPositions(state: StorageState): RevealedPosition[] {
    const cloned = serializeAndDeserialize(state);
    return [...cloned.getRevealedPositions()];
}

function stringifyRevealedPositions(positions: RevealedPosition[]): string {
    const result = JSON.stringify(positions, null, 2);
    return result;
}

describe('a storage state', () => {

    describe('that is built up', () => {
        let initialState: StorageState;

        beforeEach(() => {
            initialState = StorageState.create();
            replayRecord(initialState, winner01)
        })

        it('should serialize and deserialize correctly', () => {
            const revealed = restoreRevealedPositions(initialState);
            expect(stringifyRevealedPositions(revealed)).toMatchSnapshot();
        })

        it('should hide state', () => {
            initialState.hideState(GameState.initial.playPosition(0).playPosition(1).playPosition(3));
            const revealed = restoreRevealedPositions(initialState);
            expect(stringifyRevealedPositions(revealed)).toMatchSnapshot();
        })
    });

    describe('based on the many losers scenario', () => {
        let initialState: StorageState;

        beforeEach(() => {
            initialState = StorageState.create();
            replayRecord(initialState, manyLosers)
        })

        it('should serialize and deserialize correctly', () => {
            const revealed = restoreRevealedPositions(initialState);
            expect(stringifyRevealedPositions(revealed)).toMatchSnapshot();
        })
    })

    describe('that is empty', () => {
        let state: StorageState;

        beforeEach(() => {
            state = StorageState.create();
        })

        it('should serialize and deserialize correctly', () => {
            const revealed = restoreRevealedPositions(state);
            expect(stringifyRevealedPositions(revealed)).toMatchSnapshot();
        })
    })

    describe('that has revealed positions', () => {
        let state: StorageState;

        beforeEach(() => {
            state = StorageState.create();
            state.revealPosition({
                gameState: GameState.initial.playPosition(0).playPosition(1),
                winner: undefined
            })
        })

        it('should remove equivalent position', () => {
            state.hideState(GameState.initial.playPosition(6).playPosition(7));
            const content = [...state.getRevealedPositions()];
            expect(content).toEqual([
                {
                    gameState: GameState.initial.playPosition(0)
                }
            ])
        })

    })
})