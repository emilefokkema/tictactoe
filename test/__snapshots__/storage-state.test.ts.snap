// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`a storage state > based on the many losers scenario > should serialize and deserialize correctly 1`] = `
"[
  {
    "gameState": [
      0,
      2,
      1,
      6,
      3,
      5,
      8,
      4
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6,
        3,
        5,
        8
      ],
      "player": 2
    }
  },
  {
    "gameState": [
      0,
      2,
      1,
      6,
      3,
      5,
      8,
      7,
      4
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6,
        3,
        5,
        8,
        7
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      1,
      6,
      3,
      8
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6,
        3
      ],
      "player": 2
    }
  },
  {
    "gameState": [
      0,
      2,
      1,
      6,
      4,
      3,
      8
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6,
        4,
        3
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      1,
      6,
      4,
      5,
      8
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6,
        4,
        5
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      1,
      6,
      4,
      7,
      8
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6,
        4,
        7
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      1,
      6,
      4,
      8,
      7
    ],
    "winner": {
      "gameState": [
        0,
        2,
        1,
        6
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      1,
      6
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        1
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      4,
      6
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        4
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      5,
      6
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        5
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      6,
      4,
      1,
      5
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        6,
        4,
        1
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      6,
      4,
      5,
      8
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        6,
        4,
        5
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      6,
      4,
      7,
      8
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        6
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      6,
      4,
      8,
      5
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        6,
        4,
        8
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      7,
      6
    ],
    "winner": {
      "gameState": [
        0,
        2,
        3,
        7
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      2,
      3,
      8,
      6
    ],
    "winner": {
      "gameState": [
        0,
        2
      ],
      "player": 1
    }
  }
]"
`;

exports[`a storage state > that is built up > should hide state 1`] = `
"[
  {
    "gameState": [
      0,
      1
    ],
    "winner": {
      "gameState": [
        0,
        1
      ],
      "player": 1
    }
  }
]"
`;

exports[`a storage state > that is built up > should serialize and deserialize correctly 1`] = `
"[
  {
    "gameState": [
      0,
      1,
      3,
      2,
      6
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        2
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      4,
      6
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        4
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      5,
      6
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        5
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      6,
      4,
      2,
      5
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        6,
        4,
        2
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      6,
      4,
      5,
      8
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        6,
        4,
        5
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      6,
      4,
      7,
      5
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        6
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      6,
      4,
      8,
      5
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        6,
        4,
        8
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      7,
      6
    ],
    "winner": {
      "gameState": [
        0,
        1,
        3,
        7
      ],
      "player": 1
    }
  },
  {
    "gameState": [
      0,
      1,
      3,
      8,
      6
    ],
    "winner": {
      "gameState": [
        0,
        1
      ],
      "player": 1
    }
  }
]"
`;

exports[`a storage state > that is empty > should serialize and deserialize correctly 1`] = `"[]"`;
